docs/design/features/source-generator-pinvokes.md

```diff
@@ -0,0 +1,233 @@
+
+A consequence of this approach is that marshaling code is not immediately available post-link for AOT scenarios (e.g. [`crossgen`](../../workflow/building/coreclr/crossgen.md) and [`crossgen2`](crossgen2-compilation-structure-enhancements.md)). The immediate unavailability of this code has been mitigated by a complex mechanism to have marshalling code generated at AOT time.
+
+The user experience of the built-in generation initially appears ideal, but there are several negative consequences that make the system costly in the long term:
+
+* Bug fixes in the marshaling system require an update to the entire runtime.
+* New types require enhancements to the marshaling system for efficient marshal behavior.
+  * [`ICustomMarshaler`](https://docs.microsoft.com/dotnet/api/system.runtime.interopservices.icustommarshaler) incurs a substantial performance penalty.
+* Once a marshaling bug becomes expected behavior the bug is difficult to fix. This is due to user reliance on shipped behavior and since the marshaling system is built into the runtime there aren't ways to select previous or new behavior.
+  * Example involving COM marshaling: https://github.com/dotnet/coreclr/pull/23974.
+* Debugging the auto-generated marshaling IL Stub is difficult for runtime developers and close to impossible for consumers of P/Invokes.
+
+This is not to say the P/Invoke system should be completely redesigned. The current system is heavily used and its simplicity for consuming native assets is a benefit. Rather this new mechanism is designed to provide a way for marshaling code to be generated by an external tool but integrate with `DllImportAttribute` in a way that isn't onerous on current .NET developers.
+
+The [Roslyn Compiler](https://github.com/dotnet/roslyn) team is working on a [Source Generator feature][source_gen_link] that will allow the generation of additional source files that can be added to an assembly during the compilation process - the runtime generation IL Stubs is an in-memory version of this scenario.
+
+**Note** This proposal is targeted at addressing P/Invoke improvements but could be adapted to work with COM interop utilizing the new [`ComWrappers`][comwrappers_link] API.
+
+### Requirements
+
+* [Source generators][source_gen_link]
+  * Branch: https://github.com/dotnet/roslyn/tree/features/source-generators
```

Comment:
Should this link to https://github.com/dotnet/roslyn/blob/features/source-generators/docs/features/source-generators.md instead?

Reply:
Sure.